solidity/contracts/isms/hook/ArbL2ToL1Ism.sol

```diff
@@ -0,0 +1,144 @@
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @title ArbL2ToL1Ism
+ * @notice Uses the native Arbitrum bridge to verify interchain messages from L2 to L1.
+ */
+contract ArbL2ToL1Ism is
+    CrossChainEnabledArbitrumL1,
+    AbstractMessageIdAuthorizedIsm
+{
+    using Message for bytes;
+    // ============ Constants ============
+
+    // module type for the ISM
+    uint8 public constant moduleType =
+        uint8(IInterchainSecurityModule.Types.ARB_L2_TO_L1);
+    // arbitrum nitro contract on L1 to forward verification
+    IOutbox public arbOutbox;
+
+    // ============ Constructor ============
+
+    constructor(
+        address _bridge,
+        address _outbox
+    ) CrossChainEnabledArbitrumL1(_bridge) {
+        require(
+            Address.isContract(_bridge),
+            "ArbL2ToL1Ism: invalid Arbitrum Bridge"
+        );
+        arbOutbox = IOutbox(_outbox);
+    }
+
+    // ============ External Functions ============
+
+    /// @inheritdoc IInterchainSecurityModule
+    function verify(
+        bytes calldata metadata,
+        bytes calldata message
+    ) external override returns (bool) {
+        return
+            _statefulVerify(metadata, message) ||
+            _verifyWithOutboxCall(metadata, message);
+    }
+
+    // ============ Internal function ============
+
+    /**
+     * @notice Verify message directly using the arbOutbox.executeTransaction function.
+     * @dev This is a fallback in case the message is not verified by the stateful verify function first.
+     * @dev This function doesn't support msg.value as the ism.verify call doesn't support it either.
+     */
+    function _verifyWithOutboxCall(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal returns (bool) {
+        (
+            bytes32[] memory proof,
+            uint256 index,
+            address l2Sender,
+            address to,
+            uint256 l2Block,
+            uint256 l1Block,
+            uint256 l2Timestamp,
+            bytes memory data
+        ) = abi.decode(
+                metadata,
+                (
+                    bytes32[],
+                    uint256,
+                    address,
+                    address,
+                    uint256,
+                    uint256,
+                    uint256,
+                    bytes
+                )
+            );
+
+        // check if the sender of the l2 message is the authorized hook
+        require(
+            l2Sender == TypeCasts.bytes32ToAddress(authorizedHook),
+            "ArbL2ToL1Ism: l2Sender != authorizedHook"
+        );
+
+        bytes32 messageId = message.id();
+        {
+            // for parsing the message id from the verifyMessageId calldata
+            bytes32 convertedBytes;
+            assembly {
+                convertedBytes := mload(add(data, 36))
```

Comment:
please avoid assembly
how about this?
```solidity
(bytes4, bytes32 parsedMessageId) = abi.decode(data, (bytes4, bytes32));
```

Reply:
switched to slicing here: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3853/commits/b5a78f3edb1d48650dbea47cd7dd92117f79aa0a